pnpm: refactor fetchDeps and configHook to top-level attributes

Scrumplex opened the pull request:
This PR brings pnpm in line with other fetchers and config hooks from the Node ecosystem.

See commit descriptions for details of each change.

In the future, we will want to make `pnpm.fetchDeps` and `pnpm.configHook` throw, instead of warn. Perhaps we can do that in a few months to give the ecosystem time to update.

## Some design questions

- ~~Currently, `pnpmConfigHook` propagates `pnpm` as a build input. In order to change the version propagated by the hook, you currently need to override its input args, which is quite clunky for such a small feature. Should we just expect `pnpm` to be added to nativeBuildInputs explicitly? That way we can save on the override there and make it potentially more readable.~~
  - This has been implemented.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Scrumplex committed:
fetchPnpmDeps,pnpmConfigHook: move to top-level attribute

move pnpm.{fetchDeps,configHook} to fetchPnpmDeps and pnpmConfigHook
respectively. Also adds aliases to the former package-level attributes.

Additionally, pnpmConfigHook does not propagate pnpm anymore, to make
changing pnpm versions easier.

This brings pnpm in line with the other Node.js/JavaScript tooling in
nixpkgs.

This doesn't cause any rebuilds.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
treewide: migrate to fetchPnpmDeps and pnpmConfigHook

This moves all packages to the new top-level attributes introduced
previously.

This doesn't cause any rebuilds.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
pnpm: add warning for deprecated fetcher and configHook attributes

Deprecate package-level pnpm tooling in favor of top-level attributes.

This doesn't cause any rebuilds.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Eveeifyeve commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
Can you also in pnpm use tar? there is a lot of benefits like storage and network traffic that was mentioned that wanted to be done by @getchoo here: https://github.com/NixOS/nixpkgs/pull/376299#discussion_r1929997815

Eveeifyeve replied:
If that needs to be done in a separate pr after this one, it can.

Scrumplex replied:
Probably needs a separate PR. We could introduce this as `fetcherVersion` 3 maybe

Scrumplex commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 469437`
Commit: `ceea59b9c2cce3f3ce9c995321cbf87589aeb381`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 19 packages built:</summary>
  <ul>
    <li>claude-code-router</li>
    <li>e-search</li>
    <li>fedistar</li>
    <li>homebox</li>
    <li>immich</li>
    <li>nixpkgs-manual</li>
    <li>paperless-ngx</li>
    <li>paperless-ngx.dist</li>
    <li>pnpmConfigHook</li>
    <li>satisfactorymodmanager</li>
    <li>taler-challenger</li>
    <li>taler-mdb</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vscode-extensions.eamodio.gitlens</li>
    <li>vscode-extensions.kilocode.kilo-code</li>
    <li>vscode-extensions.rooveterinaryinc.roo-cline</li>
    <li>zipline</li>
  </ul>
</details>

gepbird commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -327,16 +327,16 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+pnpm is available as the top-level package `pnpm`. Additionally, there are variants pinned to certain major versions, like `pnpm_8`, `pnpm_9` and `pnpm_10`, which support different sets of lock file versions.
 
-When packaging an application that includes a `pnpm-lock.yaml`, you need to fetch the pnpm store for that project using a fixed-output-derivation. The functions `pnpm_8.fetchDeps` and `pnpm_9.fetchDeps` can create this pnpm store derivation. In conjunction, the setup hooks `pnpm_8.configHook` and `pnpm_9.configHook` will prepare the build environment to install the pre-fetched dependencies store. ...
+When packaging an application that includes a `pnpm-lock.yaml`, you need to fetch the pnpm store for that project using a fixed-output-derivation. The function `pnpmFetchDeps` can create this pnpm store derivation. In conjunction, the setup hook `pnpmConfigHook` will prepare the build environment to install the pre-fetched dependencies store. Here is an example for a package that contains `packag ...
```
```suggestion
When packaging an application that includes a `pnpm-lock.yaml`, you need to fetch the pnpm store for that project using a fixed-output-derivation. The function `fetchPnpmDeps` can create this pnpm store derivation. In conjunction, the setup hook `pnpmConfigHook` will prepare the build environment to install the pre-fetched dependencies store. Here is an example for a package that contains `package.json` and a `pnpm-lock.yaml` files using the fetcher and setup hook above:
```

gepbird commented:
If we decide that a `buildPnpmPackage` function should reduce the ugliness/boilerplate of using a different pnpm version, maybe it's better to postpone the treewide in a separate PR that includes both changes?

gepbird reviewed (changes requested):
Thanks for your work!
I haven't checked the first treewide yet, the rest of the changes look good.
While I'm not the happiest with the new overriding interface (diff is +1008 −418), I don't have a better idea and I'm fine with it.

Can you rebase to fix the merge conflict? After that I'll review the first treewides, rebuild all packages, all pnpm deps and if it's all good I'd say it's ready for merge.

Scrumplex committed:
fetchPnpmDeps: allow passing existing derivation args

Previously, users had to add hacks to add additional nativeBuildInputs
entries to the fetcher. This way we allow users to append
nativeBuildInputs, impureEnvVars and also merge potential passthru
attributes.

This might cause rebuilds.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

gepbird reviewed (approved):
Diff looks good. I noticed a some cases of unpinning pnpm by changing `pnpm_10` to `pnpm`, and a few cases of pinning it. I don't have problems with it, we will deal with it when `pnpm_11` gets released.

I built a few packages and probably there are no issues, but just to be sure I'll rebuild packages and pnpm deps soon.

Scrumplex commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 469437`
Commit: `ddef6d63aca5afcaa1e7634432e89ff71d20e088`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 19 packages built:</summary>
  <ul>
    <li>claude-code-router</li>
    <li>e-search</li>
    <li>fedistar</li>
    <li>homebox</li>
    <li>immich</li>
    <li>nixpkgs-manual</li>
    <li>paperless-ngx</li>
    <li>paperless-ngx.dist</li>
    <li>pnpmConfigHook</li>
    <li>satisfactorymodmanager</li>
    <li>taler-challenger</li>
    <li>taler-mdb</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vscode-extensions.eamodio.gitlens</li>
    <li>vscode-extensions.kilocode.kilo-code</li>
    <li>vscode-extensions.rooveterinaryinc.roo-cline</li>
    <li>zipline</li>
  </ul>
</details>

Scrumplex committed:
treewide: override pnpm nodejs versions

Some packages in nixpkgs were using different nodejs versions, but were
not overriding pnpm to match. Apart from reducing the derivation size,
this could prevent future issues due to version mismatches.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
e-search: add gitMinimal to fetchPnpmDeps using nativeBuildInputs

fetchPnpmDeps now allows merging attributes like nativeBuildInputs
making workarounds like these obsolete.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
vencord: combine overridden attributes for fetchPnpmDeps

`fetchPnpmDeps` accepts all kinds of attributes, so overriding them is
almost never necessary.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented on pkgs/development/tools/pnpm/generic.nix:
```diff
@@ -71,12 +72,20 @@ stdenvNoCC.mkDerivation (finalAttrs: {
+        };
+      configHook = pnpmConfigHook.overrideAttrs (prevAttrs: {
+        propagatedBuildInputs = prevAttrs.propagatedBuildInputs or [ ] ++ [
+          pnpm'
```
This is for backwards compatibility, yes. I didn't want to change the semantics of the package level attributes here, so we can backport this safely to 25.11 (except for the commit that deprecates the package attributes)

gepbird commented on pkgs/by-name/si/signal-desktop/package.nix:
```diff
@@ -66,16 +68,17 @@ let
     nativeBuildInputs = [
       nodejs
-      pnpm.configHook
+      (pnpmConfigHook.override { inherit pnpm; })
```
Can you update also this?

gepbird commented on pkgs/by-name/wr/wrangler/package.nix:
```diff
@@ -59,7 +65,7 @@ stdenv.mkDerivation (finalAttrs: {
     makeWrapper
     nodejs
-    pnpm_9.configHook
+    (pnpmConfigHook.override { inherit pnpm; })
```
Ditto

gepbird commented on pkgs/servers/authelia/default.nix:
```diff
@@ -2,13 +2,21 @@
+      pnpmConfigHook
+      fetchFromGitHub
+      ;
+  },
```
This removed `pnpm`, but seems like the package still builds.
I'm not sure why are only some of the arguments listed here while others are inferred like `stdenv`, I imagine it's because of overriding, which is probably only needed for `nodejs` and `pnpm`.

gepbird reviewed (changes requested):
Other than the below, diff looks good.

I started nixpkgs-review, will report back after ~8 hours.

gepbird reviewed (approved):
No known failures, aditionally I rebuilt some pnpm deps:
- `bash-language-server.pnpmDeps`
- `zenn-cli.pnpmDeps`
- `en-croissant.pnpmDeps`
- `it-tools.pnpmDeps`
- `modrinth-app-unwrapped.pnpmDeps`
